### PR TITLE
Standard daemonizing

### PR DESCRIPTION
**Pull request by eviljoel**

### CODE REVIEW

File: src/usr/share/torwatchdog/torwatchdog.py [hunk 1]
```diff
@@ -15,183 +15,321 @@
...
+        config_helper.verify_string_exists(config_parser, 'email_subject')
+
+    return (config, config_helper, logger)
+
+
+# Creates directories if they do not exist and sets the specified ownership and permissions.
+#
+# system_path: The system path that the directories should be created under. These are assumed to
+#   already exist. The ownership and permissions on these directories are not modified.
+# program_dirs: Additional directories that should be created under the system path that should take
```
Comment (emfrost) [inline src/usr/share/torwatchdog/torwatchdog.py]: `program_dirs` seems like a name for a list, but passing a list will throw an exception.

File: src/usr/share/torwatchdog/torwatchdog.py [hunk 2]
```diff
@@ -15,183 +15,321 @@
...
-            logger.warn("Send down notification")
-
-            message = gpgmailmessage.GpgMailMessage()
-            message.set_subject(config['subject'])
-            message.set_body('Down notification for %s at %s.' % (config['url'], datetime.datetime.now()))
-            message.queue_for_sending()
+            message = 'Down notification for %s at %s.' % (config['url'], datetime.datetime.now())
+            logger.warn(message)
+
+            try:
```
Comment (emfrost) [inline src/usr/share/torwatchdog/torwatchdog.py]: This try block is almost exactly duplicated for the next condition. Let's consider breaking it out into a function.

File: src/usr/share/torwatchdog/torwatchdog.py [hunk 3]
```diff
@@ -15,183 +15,321 @@
...
+
+    tor_process = start_tor(config)
+
+    with daemon_context:
+        main_loop(config)
+ 
+except Exception as e:
+    logger.critical('Fatal %s: %s\n' % (type(e).__name__, e.message))
+    logger.critical(traceback.format_exc())
+    if tor_process != None:
```
Comment (emfrost) [inline src/usr/share/torwatchdog/torwatchdog.py]: This block is exactly the same as `sig_term_handler`. Is there a particular reason you didn't just call that instead?

File: debian/control [hunk 4]
```diff
@@ -7,5 +7,5 @@ Standards-Version: 3.9.6.1
 
 Package: torwatchdog
 Architecture: all
-Depends: confighelper (>=0.3),gpgmailer (>=0.7),python-stem,python-socksipy,tor
+Depends: adduser (>= 3.11),confighelper (>=0.3),gpgmailer (>=0.7),python-daemon,python-stem,python-socks,tor
```
Comment (emfrost) [inline debian/control]: python-socks does not exist in the Ubuntu 14.04 repositories, and python-socksipy is not in the Ubuntu 16.04 repositories. Debhelper supports alternative package names, though.

Comment (emfrost) [inline debian/control]: python-lockfile is its own package in Ubuntu 14.04, and should be a dependency.

File: src/usr/share/torwatchdog/torwatchdog.py [hunk 5]
```diff
@@ -15,183 +15,321 @@
...
+
 import confighelper
 import ConfigParser
+import daemon
 import datetime
 import gpgmailmessage
+import grp
 import logging
 import os
+from lockfile import pidlockfile
```
Comment (emfrost) [inline src/usr/share/torwatchdog/torwatchdog.py]: In Ubuntu 14.04, pidlockfile lives in the daemon package, not lockfile.

**Review by emfrost — CHANGES_REQUESTED** [review]
The package does not install in the currently supported distribution. Reverting the distro-specific breaks allows it to run as intended. It runs as a non-root user and creates files with the correct permissions.

File: src/usr/share/torwatchdog/torwatchdog.py (hunk 1 above)
Comment (eviljoel) [reply src/usr/share/torwatchdog/torwatchdog.py]: I'll guess I'll make the comment clearer.

File: src/usr/share/torwatchdog/torwatchdog.py (hunk 2 above)
Comment (eviljoel) [reply src/usr/share/torwatchdog/torwatchdog.py]: OK.

File: src/usr/share/torwatchdog/torwatchdog.py (hunk 3 above)
Comment (eviljoel) [reply src/usr/share/torwatchdog/torwatchdog.py]: What am I going to put for the stack frame parameter? It doesn't quite fit. Also, I don't think 3 lines is justification enough to break out into its own method.

File: debian/control (hunk 4 above)
Comment (eviljoel) [reply debian/control]: It is a dependency of python-daemon. No need to include it here explicitly.

Comment (eviljoel) [reply debian/control]: Added alternative package name.

File: src/usr/share/torwatchdog/torwatchdog.py (hunk 5 above)
Comment (eviljoel) [reply src/usr/share/torwatchdog/torwatchdog.py]: Added fallback import.

**Review by emfrost — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -1,47 +1,65 @@
...
-3.   Use `debuild` in the project root directory to build the package.
-4.   Use `dpkg -i` to install the package.
-5.   Use `apt-get -f install` to resolve any missing dependencies. The daemon will attempt to start and fail. (This is expected.)
-6.   Locate the example configuration file at `/etc/torwatchdog/torwatchdog.conf.example`. Copy or rename this file to `torwatchdog.conf` in the same directory. Edit this file to enter the Tor URL you want to monitor or to modify other settings.
-7.   Restart the daemon with `service torwatchdog restart`. If the configuration file is valid and named correctly, the service will now start successfully.
+1.  Manually add the Tor Project's Apt repository to ensure you have a secure and updated
+    version of Tor. Instructions located here:
+    https://www.torproject.org/docs/debian.html.en
+2.  Clone the repository and checkout the lastest release tag. (Do not build against the
+    `master` branch. The `master` branch might not be stable.)
```
Comment (emfrost) [inline README.md:46]: GpgMailer uses "may not be stable" in this line. Does this inconsistency matter?

File: src/usr/share/torwatchdog/torwatchdog.py [hunk 7]
```diff
@@ -15,174 +15,226 @@
...
+    """Configures the tor proxy settings.
+
+    config: The program configuration object, mostly based on the configuration file.
+    """
     # Set socks proxy and wrap the urllib module
-    # TODO: Consider choosing a randomly available TCP port.
-    socks.setdefaultproxy(socks.PROXY_TYPE_SOCKS5, '127.0.0.1', int(config['tor_socks_port']))
+    # TODO: Eventually consider choosing a randomly available TCP port.
+    socks.setdefaultproxy(
+        socks.PROXY_TYPE_SOCKS5, '127.0.0.1', int(config['tor_socks_port']))
```
Comment (emfrost) [inline src/usr/share/torwatchdog/torwatchdog.py]: `config['tor_socks_port']` is already definitely an int, no need to typecast it again.

File: README.md [hunk 8]
```diff
@@ -1,47 +1,65 @@
...
+3.  Use `debuild` in the project root directory to build the package.
+4.  Use `dpkg -i` to install the package.
+5.  Run `apt-get -f install` to resolve any missing dependencies. The daemon will attempt to
+    start and fail. (This is expected.)
+6.  Copy or rename the example configuration file
+    `/etc/torwatchdog/torwatchdog.conf.example` to `/etc/torwatchdog/torwatchdog.conf`. Edit
+    this file to enter the Tor URL you want to monitor. Other settings can also be modified.
+7.  Use `chmod` to clear the _other user_ permissions bits of `watchman.conf`. Namely, remove
+    read, write, and execute permissions for _other_.
+8.  To ease system maintenance, add `torwatchdog` as a supplimental group to administrative
```
Comment (emfrost) [inline README.md]: "Supplimental" is spelled supplemental.

File: src/usr/share/torwatchdog/torwatchdog.py [hunk 9]
```diff
@@ -15,174 +15,226 @@
...
+    # Temporarily drop permissions and create the handle to the logger.
     os.setegid(program_gid)
     os.seteuid(program_uid)
-    config_helper.configure_logger(os.path.join(log_dir, log_file), config['log_level'])
-    os.seteuid(os.getuid())
-    os.setegid(os.getgid())
+    config_helper.configure_logger(os.path.join(LOG_DIR, LOG_FILE), config['log_level'])
 
-    logger = logging.getLogger('%s-daemon' % program_name)
+    logger = logging.getLogger('%s-daemon' % PROGRAM_NAME)
```
Comment (emfrost) [inline src/usr/share/torwatchdog/torwatchdog.py]: The naming convention for this is `getLogger(__name__)`.

File: README.md (hunk 8 above)
Comment (eviljoel) [reply README.md]: This was also a problem in the other projects.

**Pull request merged**